coverage_comment/coverage.py

```diff
@@ -211,57 +217,94 @@ def extract_info(data: dict, coverage_path: pathlib.Path) -> Coverage:
     )
 
 
-def get_diff_coverage_info(base_ref: str, coverage_path: pathlib.Path) -> DiffCoverage:
-    subprocess.run("git", "fetch", "--depth=1000", path=pathlib.Path.cwd())
-    subprocess.run("coverage", "xml", path=coverage_path)
-    with tempfile.NamedTemporaryFile("r") as f:
-        subprocess.run(
-            "diff-cover",
-            "coverage.xml",
-            f"--compare-branch=origin/{base_ref}",
-            f"--json-report={f.name}",
-            "--diff-range-notation=..",
-            "--quiet",
-            path=coverage_path,
+def get_diff_coverage_info(
+    added_lines: dict[pathlib.Path, list[int]], coverage: Coverage
+) -> DiffCoverage:
+    files = {}
+    total_num_lines = 0
+    total_num_violations = 0
+    num_changed_lines = 0
+
+    for path, added_lines_for_file in added_lines.items():
+        num_changed_lines += len(added_lines_for_file)
+
+        try:
+            file = coverage.files[path]
+        except KeyError:
+            continue
+
+        executed = set(file.executed_lines) & set(added_lines_for_file)
+        count_executed = len(executed)
+
+        missing = set(file.missing_lines) & set(added_lines_for_file)
+        count_missing = len(missing)
+        # We don't want to count twice the lines that are both executed and
+        # missing (e.g. the branch lines partially executed)
+        count_total = len(executed | missing)
```

Comment:
I may well be wrong about that, but I don't think this is necessary. To my knowledge lines are never in both `executed_lines` and `missing_lines` and they indicate the branches [via `executed_branches` and `missing_branches`](https://github.com/nedbat/coveragepy/blob/afa75596342d2463a5676486885d1b0b489e20c0/coverage/jsonreport.py#L114-L119).

Reply:
You're probably right, but I guess the easiest way to find out is to test :)

Reply:
You're right :)
```console
$ tree
.
├── coverage.json
├── my.py
└── test_x.py

1 directory, 3 files

$ cat my.py
import os


def x():
    if os.environ.get("A"):
        print("A")
    print("B")

$ cat test_x.py
from my import x


def test_x():
    x()

$ pytest --cov-report=json --cov my --cov-branch
====================================================== test session starts =======================================================
platform darwin -- Python 3.9.18, pytest-7.4.2, pluggy-1.3.0
rootdir: /private/tmp/plop
plugins: cov-4.1.0
collected 1 item

test_x.py .                                                                                                                [100%]

---------- coverage: platform darwin, python 3.9.18-final-0 ----------
Coverage JSON written to file coverage.json


======================================================= 1 passed in 0.12s ========================================================

$ cat coverage.json | jq '.files."my.py" | .executed_lines, .missing_lines'
[
  1,
  4,
  5,
  7
]
[
  6
]
```